modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java

```diff
@@ -7,83 +7,110 @@
+        final Optional<MetadataKeyIdModelProperty> keyIdModelProperty = component.getModelProperty(MetadataKeyIdModelProperty.class);
 
+        Object keyIdObject;
 
-        Class<?> typeClass = getType(type);
-        checkInstantiable(typeClass);
-        fillFieldValueMap(key, typeClass);
+        if (isKeyLessComponent(metadataKeyParts))
+        {
+            keyIdObject = new NullMetadataKey().getId();
+        }
+        else
+        {
+            checkKeyIdModelIsPresent(component, keyIdModelProperty);
+            keyIdObject = resolveMetadataKeyWhenPresent(key, metadataKeyParts, keyIdModelProperty.get());
+        }
 
+        return keyIdObject;
     }
 
-    private void fillFieldValueMap(MetadataKey key, Class<?> type) throws MetadataResolvingException
+    private static Object resolveMetadataKeyWhenPresent(MetadataKey key, List<ParameterModel> metadataKeyParts, MetadataKeyIdModelProperty keyIdModelProperty) throws MetadataResolvingException
     {
-        Map<Integer, Field> metadataKeyPartMap = getAnnotatedFields(type, MetadataKeyPart.class)
-                .stream().collect(new ImmutableMapCollector<>(f -> f.getAnnotation(MetadataKeyPart.class).order(), identity()));
+        Object keyIdObject;
 
-        int order = 1;
-        while (!key.getChilds().isEmpty())
+        if (isSingleLevelKey(metadataKeyParts))
         {
-            fieldValueMap.put(metadataKeyPartMap.get(order), key.getId());
-            key = key.getChilds().iterator().next();
-            order++;
+            return key.getId();
         }
-        fieldValueMap.put(metadataKeyPartMap.get(order), key.getId());
+        else
+        {
+            keyIdObject = resolveMultiLevelKey(keyIdModelProperty, key);
+        }
+        return keyIdObject;
     }
 
     /**
-     * @return a new instance of the {@link MetadataKeyId} parameter {@code type} with the values of the
-     * passed {@link MetadataKey}.
-     * @throws MetadataResolvingException if the instantiation fails.
+     * Resolves the KeyIdObject for a MultiLevel {@link MetadataKeyId}
+     *
+     * @param keyIdModelProperty model property of the {@link MetadataKeyId} parameter
+     * @param key                key containing the values of each level
+     * @return the KeyIdObject for the {@link MetadataKeyId} parameter
+     * @throws MetadataResolvingException
      */
-    public Object resolve() throws MetadataResolvingException
+    private static Object resolveMultiLevelKey(MetadataKeyIdModelProperty keyIdModelProperty, MetadataKey key) throws MetadataResolvingException
     {
+        final Class<?> type = getType(keyIdModelProperty.getType());
+        checkInstantiable(type);
+
+        final Map<Field, String> fieldValueMap = fillFieldValueMap(type, key);
+
         Object metadataKeyId;
         try
         {
-            metadataKeyId = getType(type).newInstance();
+            metadataKeyId = type.newInstance();
         }
         catch (Exception e)
         {
```

Comment:
It should not be a INVALID_METADATA_KEY code, due that is not that the given MetadataKey is invalid, the created extension is invalid. 
If this happen, is that was a design time error that the MetadataKeyId was not validated correctly, and giving to the metadata consumer the error of invalid metadata key is confusing